content/exsaml.xqm

```diff
@@ -367,18 +378,23 @@ declare %private function exsaml:validate-saml-assertion($assertion as item()) {
 };
 
 (: retrieve issued SAML request id and delete if answered :)
+(: this is pointless if we are configured to accept unsolicited SAML assertions :)
 declare %private function exsaml:check-authnreqid($reqid as xs:string) {
-    let $log := exsaml:log("info", "verifying SAML request id: " || $reqid)
-    return
-        if (system:as-user($exsaml:exsaml-user, $exsaml:exsaml-pass,
-                exists(doc($exsaml:saml-coll-reqid||"/"||$reqid)) and empty(xmldb:remove($exsaml:saml-coll-reqid, $reqid)))) then
-            $reqid
-        else ""
+    if (not($exsaml:idp-unsolicited)) then (
+        let $log := exsaml:log("info", "verifying SAML request id: " || $reqid)
+        let $saml-coll-reqid := $exsaml:saml-coll-reqid-base || "/" || $exsaml:saml-coll-reqid-name
+        return
+            if (system:as-user($exsaml:exsaml-user,
+                               $exsaml:exsaml-pass,
+                               exists(doc($saml-coll-reqid||"/"||$reqid))
+                and empty(xmldb:remove($saml-coll-reqid, $reqid)))) 
```

Comment:
There is a slight possibility that xmldb:remove is not called. Because existdb might not evaluate the right hand side of an `and` if the left hand side evaluates to `false()`.

Reply:
this is on purpose. no need to remove if it does not exist